src/D2L.CodeStyle.Analyzers/Immutability/TypeDeclarationImmutabilityAnalyzer.cs

```diff
@@ -0,0 +1,82 @@
+﻿using System.Collections.Immutable;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace D2L.CodeStyle.Analyzers.Immutability {
+	[DiagnosticAnalyzer( LanguageNames.CSharp )]
+	public sealed class TypeDeclarationImmutabilityAnalyzer : DiagnosticAnalyzer {
+
+		public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(
+			Diagnostics.ArraysAreMutable,
+			Diagnostics.DelegateTypesPossiblyMutable,
+			Diagnostics.DynamicObjectsAreMutable,
+			Diagnostics.EventMemberMutable,
+			Diagnostics.MemberIsNotReadOnly,
+			Diagnostics.NonImmutableTypeHeldByImmutable,
+			Diagnostics.TypeParameterIsNotKnownToBeImmutable,
+			Diagnostics.UnexpectedMemberKind,
+			Diagnostics.UnexpectedTypeKind,
+			Diagnostics.UnnecessaryMutabilityAnnotation
+		);
+
+		public override void Initialize( AnalysisContext context ) {
+			context.EnableConcurrentExecution();
+			context.RegisterCompilationStartAction( CompilationStart );
+		}
+
+		public static void CompilationStart(
+			CompilationStartAnalysisContext context
+		) {
+			ImmutabilityContext immutabilityContext = ImmutabilityContext.Create( context.Compilation );
+
+			context.RegisterSymbolAction(

```

Comment:
Does this only run once if we have two decls for a partial class?

Reply:
(If it runs twice and emits the same diagnostics I wouldn't be surprised if they were de-duped, so check carefully)

Reply:
Can double check, but no I don't think so. Symbols are a different layer than syntax, so I would be very surprised.

Reply:
Yeah, the docs say:

> Register an action to be executed at completion of semantic analysis of an ISymbol with an appropriate Kind. A symbol action reports Diagnostics about ISymbols.

but I don't know enough to say for sure based on that, personally. I know semantic models are somewhat tied to documents and/or SyntaxTrees which is where my concern comes from.

Reply:
Like worst case we live with it or we add a cache, whatever.